Code/GraphMol/Sgroup.cpp

```diff
@@ -0,0 +1,322 @@
+//
+//
+//  Copyright (C) 2002-2018 Greg Landrum and T5 Informatics GmbH
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+#include "Sgroup.h"
+#include "ROMol.h"
+
+namespace RDKit {
+
+static const std::vector<std::string> sGroupTypes = {
+    // polymer sgroups:
+    "SRU", "MON", "COP", "CRO", "GRA", "MOD", "MER", "ANY",
+    // formulations/mixtures:
+    "COM", "MIX", "FOR",
+    // other
+    "SUP", "MUL", "DAT", "GEN"};
+
+static const std::vector<std::string> sGroupSubtypes = {"ALT", "RAN", "BLO"};
+static const std::vector<std::string> sGroupConnectTypes = {"HH", "HT", "EU"};
+
+SGroup::SGroup(ROMol *owning_mol, const std::string &_type)
+    : d_id(0), d_type(_type), dp_mol(owning_mol) {
+  PRECONDITION(owning_mol, "owning molecule is nullptr");
+}
+
+void SGroup::setOwningMol(ROMol *mol) {
+  PRECONDITION(mol, "owning molecule is nullptr");
+
+  dp_mol = mol;
+}
+
+const std::string &SGroup::getProp(const std::string &prop) const {
+  try {
+    return d_prop.at(prop);
+  } catch (const std::out_of_range &) {
+    std::ostringstream errout;
+    errout << "String Property '" << prop << "' not found in SGroup "
+           << getId();
+    throw SGroupException(errout.str());
+  }
+}
+
+//! check if SGroup has the given property set
+bool SGroup::hasProp(const std::string &prop) const {
+  return d_prop.find(prop) != d_prop.end();
+}
+
+namespace {
+bool isSGroupIdFree(const ROMol &mol, unsigned int id) {
+  auto sgroups = getSGroups(mol);
+  CHECK_INVARIANT(sgroups, "no sgroups on molecule owning an sgroup");
+  auto match_sgroup = [&](const SGROUP_SPTR &sg) { return id == sg->getId(); };
+  return std::find_if(sgroups->begin(), sgroups->end(), match_sgroup) ==
+         sgroups->end();
+}
+}  // namespace
+
+void SGroup::setId(unsigned int id) {
+  PRECONDITION(dp_mol, "SGroup is not owned by any molecule");
```

Comment:
You can't construct an sgroup without an owning mol, so the PRECONDITION appears to be unnecessary.